github.com/cockroachdb/pebble/internal/metamorphic: TestMetaTwoInstance failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/13606093?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/13606093?buildTab=artifacts#meta) on refs/heads/master @ [d4e335572f1a](https://github.com/cockroachdb/pebble/commits/d4e335572f1a):


```
        iter21.SeekLT("yltmchb@7", "") // [false] <nil> #2014
        iter21.SetBounds("yltmchb@7", "yltmchb@7") // <nil> #2017
        iter21.SeekGE("yltmchb@7", "") // [false] <nil> #2018
        iter21.Prev("") // [false] <nil> #2019
        iter21.Prev("") // [false] <nil> #2020
        iter21.Next("") // [false] <nil> #2021
        iter21.SeekGE("abzynybhk@8", "") // [false] <nil> #2025
        iter21.SetBounds("zbylcpqfvtd@2", "zkgoitppuh@5") // <nil> #2026
        iter21.SeekGE("zbylcpqfvtd@2", "") // [true,"zbylcpqfvtd@2",<no point>,["zbylcpqfvtd@2","zejql")=>{"@5"="edeobljcubuozn"}*] <nil> #2027
        // FATAL: levelIter L6: lower bound violation: zbylcpqfvtd@7#0,1 < zbylcpqfvtd@2
        // FATAL: goroutine 61 [running]:
        // FATAL: runtime/debug.Stack()
        // FATAL: 	GOROOT/src/runtime/debug/stack.go:24 +0x5e
        // FATAL: github.com/cockroachdb/pebble.(*levelIter).verify(0xc000654a40, 0xc000414700, {{0x7f7cdc009f23?, 0x13?, 0x2?}, 0x0?})
        // FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/level_iter.go:719 +0xd7
        // FATAL: github.com/cockroachdb/pebble.(*levelIter).SeekLT(0xc000654a40, {0xc00015b4ad, 0xc, 0x13}, 0xd?)
        // FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/level_iter.go:840 +0xe5
        // FATAL: github.com/cockroachdb/pebble/internal/invalidating.(*iter).SeekLT(0xc000566240, {0xc00015b4ad?, 0xc?, 0x13?}, 0xd8?)
        // FATAL: 	github.com/cockroachdb/pebble/internal/invalidating/external/com_github_cockroachdb_pebble/internal/invalidating/iter.go:124 +0x28
        // FATAL: github.com/cockroachdb/pebble.(*mergingIter).switchToMaxHeap(0xc000653df8)
        // FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/merging_iter.go:570 +0x1f0
        // FATAL: github.com/cockroachdb/pebble.(*mergingIter).Prev(0xc000653df8)
        // FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/merging_iter.go:1454 +0x8a
        // FATAL: github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).Prev(0xc000658628)
        // FATAL: 	github.com/cockroachdb/pebble/internal/keyspan/external/com_github_cockroachdb_pebble/internal/keyspan/interleaving_iter.go:544 +0x10e
        // FATAL: github.com/cockroachdb/pebble.(*Iterator).prevUserKey(0xc000653800)
        // FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:1129 +0x148
        // FATAL: github.com/cockroachdb/pebble.(*Iterator).PrevWithLimit(0xc000653800, {0x0, 0x0, 0x0})
        // FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:2015 +0x1e7
        // FATAL: github.com/cockroachdb/pebble.(*Iterator).Prev(...)
        // FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:1931
        // FATAL: github.com/cockroachdb/pebble/metamorphic.(*retryableIter).Prev.func1()
        // FATAL: 	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/retryable.go:132 +0x2a
        // FATAL: github.com/cockroachdb/pebble/metamorphic.(*retryableIter).withRetry(0xc0006da7e0, 0xc000887db0)
        // FATAL: 	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/retryable.go:47 +0x2b
        // FATAL: github.com/cockroachdb/pebble/metamorphic.(*retryableIter).Prev(0xc0003952a0?)
        // FATAL: 	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/retryable.go:131 +0x47
        // FATAL: github.com/cockroachdb/pebble/metamorphic.(*iterPrevOp).run(0xc0002fe210, 0xc00023e0f0?, {0xc0003952a0?, 0x0?, 0x0?})
        // FATAL: 	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/ops.go:1625 +0xb4
        // FATAL: github.com/cockroachdb/pebble/metamorphic.Execute.func1()
        // FATAL: 	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/meta.go:577 +0x12d
        // FATAL: golang.org/x/sync/errgroup.(*Group).Go.func1()
        // FATAL: 	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x56
        // FATAL: created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 21
        // FATAL: 	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0x96
        iter21.Prev("") // [false] <nil> #2028
        batch16.Get("ofikmb@6") // ["koppktuo"] <nil> #2015
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMetaTwoInstance$' --run-dir /artifacts/meta/240119-064432.5072795294643/random-025 --try-to-reduce -v
        --- FAIL: TestMetaTwoInstance/execution/random-025 (0.29s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMetaTwoInstance$' ./internal/metamorphic -seed 1705646672507749385 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/13629045?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/13629045?buildTab=artifacts#meta) on refs/heads/master @ [5b092519dc3f](https://github.com/cockroachdb/pebble/commits/5b092519dc3f):


```
        db2.DeleteRange("qhyvapvkasv@4", "ydptxuctunp@6") // <nil> #3973
        iter47 = db1.NewIter("", "cabrrwu@7", 2 /* key types */, 0, 0, true /* use L6 filters */, "@8" /* masking suffix */) // <nil> #3974
        db2.Get("bwwplie@7") // [""] pebble: not found #3976
        db2.Get("fffa@10") // [""] pebble: not found #3980
        iter47.SeekLT("azesvmwntvw@7", "") // [true,"azesvmwntvw@14","fzdhmafwvfiix",<no range>] <nil> #3975
        iter47.NextPrefix() // [false] NextPrefix not permitted with upper bound cabrrwu@7 #3977
        iter47.Prev("") // [false] NextPrefix not permitted with upper bound cabrrwu@7 #3978
        iter47.InternalNext() // NextPrefix not permitted with upper bound cabrrwu@7 #3979
        iter47.SeekGE("daurchhmury@12", "zckpbhp@13") // [invalid] <nil> #3982
        iter47.SetBounds("qabhcsizc@13", "srwt@5") // <nil> #3983
        snap30.Get("pozhnmgmldk@8") // [""] pebble: not found #3984
        db2.Delete("effefunvgf@10") // <nil> #3986
        db1.Set("fffa@7", "tbieojdfj") // <nil> #3995
        snap30.Get("pjyicoastl") // [""] pebble: not found #3989
        iter47.SetBounds("cdjpop", "itcvdiaupjk@10") // <nil> #3990
        iter47.SeekLT("itcvdiaupjk@10", "") // [true,"irhm@8","j",["elzwgj","itcvdiaupjk@10")=>{"@5"="hvekbgubufrfkk"}*] <nil> #3991
        iter47.Next("") // [false] <nil> #3992
        iter47.Next("") // [false] <nil> #3993
        iter47.Prev("") // [true,"irhm@8","j",["elzwgj","itcvdiaupjk@10")=>{"@5"="hvekbgubufrfkk"}*] <nil> #3994
        iter47.SeekGE("vtvf@1", "") // [false] <nil> #4001
        iter47.Last() // [true,"irhm@8","j",["elzwgj","itcvdiaupjk@10")=>{"@5"="hvekbgubufrfkk"}*] <nil> #4002
        iter47.Next("sfztapibb@5") // [invalid] <nil> #4003
        iter47.SeekGE("eteid@8", "") // [true,"eteid@8",<no point>,["elzwgj","itcvdiaupjk@10")=>{"@5"="hvekbgubufrfkk"}*] <nil> #4005
        iter47.Last() // [true,"irhm@8","j",["elzwgj","itcvdiaupjk@10")=>{"@5"="hvekbgubufrfkk"}] <nil> #4007
        iter47.Prev("") // [true,"iajv@8","",["elzwgj","itcvdiaupjk@10")=>{"@5"="hvekbgubufrfkk"}] <nil> #4008
        iter47.SeekGE("wvntppdjg@6", "") // [false] <nil> #4009
        iter47.Next("cnhn@9") // [invalid] <nil> #4012
        iter47.SeekLT("stnjshmvu@11", "") // [true,"irhm@8","j",["elzwgj","itcvdiaupjk@10")=>{"@5"="hvekbgubufrfkk"}*] <nil> #4014
        iter47.SetBounds("cdjpop", "cdjpop") // <nil> #4016
        iter47.SeekGE("mnal@11", "") // [false] <nil> #4017
        iter47.Next("") // [false] <nil> #4018
        iter47.Next("") // [false] <nil> #4026
        iter47.SetOptions("", "cdjpop", 0 /* key types */, 0, 0, true /* use L6 filters */, "" /* masking suffix */) // <nil> #4027
        iter47.First() // [true,"azesvmwntvw@14","fzdhmafwvfiix",<no range>] <nil> #4028
        iter47.SeekLT("httnwhl@7", "") // [true,"bkofewyf@8","q",<no range>] <nil> #4029
        iter47.SeekPrefixGE("bzwevm@3") // [false] <nil> #4030
        iter47.First() // [true,"azesvmwntvw@14","fzdhmafwvfiix",<no range>] <nil> #4032
        db1.RangeKeySet("mvnptfo", "wboxjyooesf", "@12", "fismieajaacf") // <nil> #3996
        db1.Set("zrhmciwxfp@8", "tmzilyzjqvjvbzvciq") // <nil> #4006
        db2.Set("hzwpfkcrn@9", "thbktruakwdablw") // <nil> #4011
        db1.DeleteRange("njkaxrmhha@8", "svdmfslm@1") // <nil> #4031
        db2.Get("wboxjyooesf@13") // [""] pebble: not found #4035
        db1.DeleteRange("bwwplie@7", "dathyxeeqw@10") // <nil> #4036
        db2.Get("zqvtcbyo") // [""] pebble: not found #4037
        snap30.Get("wusbzwp@4") // [""] pebble: not found #4041
        iter47.Prev("") // [false] <nil> #4042
        iter47.SetBounds("flfubwrx@1", "wboxjyooesf@13") // <nil> #4048
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMetaTwoInstance$' --run-dir /artifacts/meta/240121-061937.2911827098097/random-004 --try-to-reduce -v
        --- FAIL: TestMetaTwoInstance/execution/random-004 (0.70s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMetaTwoInstance$' ./internal/metamorphic -seed 1705817977291934308 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/13656680?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/13656680?buildTab=artifacts#meta) on refs/heads/master @ [5b092519dc3f](https://github.com/cockroachdb/pebble/commits/5b092519dc3f):


```
        // FATAL: github.com/cockroachdb/pebble.(*levelIter).Prev(0xc0002a7cf0)
        // FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/level_iter.go:1000 +0x10d
        // FATAL: github.com/cockroachdb/pebble.(*mergingIter).switchToMaxHeap(0xc0002a6df8)
        // FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/merging_iter.go:580 +0x60e
        // FATAL: github.com/cockroachdb/pebble.(*mergingIter).Prev(0xc0002a6df8)
        // FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/merging_iter.go:1454 +0x8a
        // FATAL: github.com/cockroachdb/pebble/internal/invalidating.(*iter).Prev(0xc0004889c0)
        // FATAL: 	github.com/cockroachdb/pebble/internal/invalidating/external/com_github_cockroachdb_pebble/internal/invalidating/iter.go:140 +0x23
        // FATAL: github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).prevPos(0xc0001aa028)
        // FATAL: 	github.com/cockroachdb/pebble/internal/keyspan/external/com_github_cockroachdb_pebble/internal/keyspan/interleaving_iter.go:715 +0x8f
        // FATAL: github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).Prev(0xc0001aa028)
        // FATAL: 	github.com/cockroachdb/pebble/internal/keyspan/external/com_github_cockroachdb_pebble/internal/keyspan/interleaving_iter.go:575 +0x159
        // FATAL: github.com/cockroachdb/pebble.(*Iterator).prevUserKey(0xc0002a6800)
        // FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:1129 +0x148
        // FATAL: github.com/cockroachdb/pebble.(*Iterator).PrevWithLimit(0xc0002a6800, {0x0, 0x0, 0x0})
        // FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:2015 +0x1e7
        // FATAL: github.com/cockroachdb/pebble.(*Iterator).Prev(...)
        // FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:1931
        // FATAL: github.com/cockroachdb/pebble/metamorphic.(*retryableIter).Prev.func1()
        // FATAL: 	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/retryable.go:132 +0x2a
        // FATAL: github.com/cockroachdb/pebble/metamorphic.(*retryableIter).withRetry(0xc0005a0540, 0xc000a49ac8)
        // FATAL: 	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/retryable.go:47 +0x2b
        // FATAL: github.com/cockroachdb/pebble/metamorphic.(*retryableIter).Prev(0xc000335ac0?)
        // FATAL: 	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/retryable.go:131 +0x47
        // FATAL: github.com/cockroachdb/pebble/metamorphic.(*iterPrevOp).run(0xc0002377a0, 0xc0001640f0?, {0xc000335ac0?, 0xc0001d9c00?, 0x0?})
        // FATAL: 	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/ops.go:1625 +0xb4
        // FATAL: github.com/cockroachdb/pebble/metamorphic.(*Test).step(0xc0001640f0, 0xc0001d9c48?, 0xa6174f?)
        // FATAL: 	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/test.go:352 +0x48
        // FATAL: github.com/cockroachdb/pebble/metamorphic.Execute(0xc0001640f0)
        // FATAL: 	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/meta.go:542 +0x10f
        // FATAL: github.com/cockroachdb/pebble/metamorphic.RunOnce({0xe8ef20?, 0xc0001249c0}, {0x7ffdf9da88da, 0x36}, 0x0, {0xc000155180, 0x3e}, {0xc000145480, 0x3, 0x4})
        // FATAL: 	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/meta.go:526 +0xe32
        // FATAL: github.com/cockroachdb/pebble/internal/metamorphic.runTestMeta(0xc000076760?, 0x1)
        // FATAL: 	github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/meta_test.go:87 +0x179
        // FATAL: github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance(0x0?)
        // FATAL: 	github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/meta_test.go:58 +0x18
        // FATAL: testing.tRunner(0xc0001249c0, 0xdca858)
        // FATAL: 	GOROOT/src/testing/testing.go:1595 +0xff
        // FATAL: created by testing.(*T).Run in goroutine 1
        // FATAL: 	GOROOT/src/testing/testing.go:1648 +0x3ad
        iter7.Prev("") // [false] <nil> #940
        // INFO: [JOB 20] flushed 1 ingested flushable L0:000019 (943B) in 0.0s (0.0s total), output rate 30MB/s
        // INFO: [JOB 14] flushing 1 memtable (747B) to L0
        // INFO: [JOB 14] flushing: sstable created 000024
        // INFO: [JOB 14] flushing: sstable created 000025
        // INFO: [JOB 14] flushing: sstable created 000026
        // INFO: [JOB 14] flushed 1 memtable (747B) to L0 [000024 000025 000026] (3.3KB), in 0.0s (0.0s total), output rate 8.7MB/s
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMetaTwoInstance$' --run-dir /artifacts/meta/240123-072426.2311248347183/random-001 --try-to-reduce -v
        --- FAIL: TestMetaTwoInstance/execution/random-001 (0.60s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMetaTwoInstance$' ./internal/metamorphic -seed 1705994666231122931 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/ingest_test.go
+++ b/ingest_test.go
@@ -583,15 +583,23 @@ func TestExcise(t *testing.T) {
 	var mem vfs.FS
 	var d *DB
 	var flushed bool
+	var efos map[string]*EventuallyFileOnlySnapshot
 	defer func() {
+		for _, e := range efos {
+			require.NoError(t, e.Close())
+		}
 		require.NoError(t, d.Close())
 	}()
 
 	var opts *Options
 	reset := func(blockSize int) {
+		for _, e := range efos {
+			require.NoError(t, e.Close())
+		}
 		if d != nil {
 			require.NoError(t, d.Close())
 		}
+		efos = make(map[string]*EventuallyFileOnlySnapshot)
 
 		mem = vfs.NewMem()
 		require.NoError(t, mem.MkdirAll("ext", 0755))
@@ -699,6 +707,25 @@ func TestExcise(t *testing.T) {
 			}
 			return ""
 
+		case "file-only-snapshot":
+			if len(td.CmdArgs) != 1 {
+				panic("insufficient args for file-only-snapshot command")
+			}
+			name := td.CmdArgs[0].Key
+			var keyRanges []KeyRange
+			for _, line := range strings.Split(td.Input, "\n") {
+				fields := strings.Fields(line)
+				if len(fields) != 2 {
+					return "expected two fields for file-only snapshot KeyRanges"
+				}
+				kr := KeyRange{Start: []byte(fields[0]), End: []byte(fields[1])}
+				keyRanges = append(keyRanges, kr)
+			}
+
+			s := d.NewEventuallyFileOnlySnapshot(keyRanges)
+			efos[name] = s
+			return "ok"
+
 		case "get":
 			return runGetCmd(t, td, d)
 
@@ -706,6 +733,7 @@ func TestExcise(t *testing.T) {
 			opts := &IterOptions{
 				KeyTypes: IterKeyTypePointsAndRanges,
 			}
+			var reader Reader = d
 			for i := range td.CmdArgs {
 				switch td.CmdArgs[i].Key {
 				case "range-key-masking":
@@ -715,11 +743,25 @@ func TestExcise(t *testing.T) {
 							return sstable.NewTestKeysMaskingFilter()
 						},
 					}
+				case "lower":
+					if len(td.CmdArgs[i].Vals) != 1 {
+						return fmt.Sprintf(
+							"%s expects at most 1 value for lower", td.Cmd)
+					}
+					opts.LowerBound = []byte(td.CmdArgs[i].Vals[0])
+				case "upper":
+					if len(td.CmdArgs[i].Vals) != 1 {
+						return fmt.Sprintf(
+							"%s expects at most 1 value for upper", td.Cmd)
+					}
+					opts.UpperBound = []byte(td.CmdArgs[i].Vals[0])
+				case "snapshot":
+					reader = efos[td.CmdArgs[i].Vals[0]]
 				default:
 					return fmt.Sprintf("unexpected argument: %s", td.CmdArgs[i].Key)
 				}
 			}
-			iter, _ := d.NewIter(opts)
+			iter, _ := reader.NewIter(opts)
 			return runIterCmd(td, iter, true)
 
 		case "lsm":

--- a/sstable/reader_iter_single_lvl.go
+++ b/sstable/reader_iter_single_lvl.go
@@ -919,6 +919,11 @@ func (i *singleLevelIterator) virtualLastSeekLE(key []byte) (*InternalKey, base.
 		return nil, base.LazyValue{}
 	}
 	if result == loadBlockIrrelevant {
+		// Enforce the lower bound here, as we could have gone past it.
+		if i.lower != nil && i.cmp(ikey.UserKey, i.lower) < 0 {
+			i.exhaustedBounds = -1
+			return nil, base.LazyValue{}
+		}
 		// Want to skip to the previous block.
 		return i.skipBackward()
 	}
@@ -931,6 +936,11 @@ func (i *singleLevelIterator) virtualLastSeekLE(key []byte) (*InternalKey, base.
 	}
 	ikey, val = i.data.Prev()
 	if ikey != nil {
+		// Enforce the lower bound here, as we could have gone past it.
+		if i.blockLower != nil && i.cmp(ikey.UserKey, i.blockLower) < 0 {
+			i.exhaustedBounds = -1
+			return nil, base.LazyValue{}
+		}
 		return ikey, val
 	}
 	return i.skipBackward()
@@ -953,7 +963,7 @@ func (i *singleLevelIterator) SeekLT(
 		// first internal key with user key < key.
 		if cmp > 0 {
 			// Return the last key in the virtual sstable.
-			return i.virtualLast()
+			return i.maybeVerifyKey(i.virtualLast())
 		}
 	}
 
@@ -1133,7 +1143,7 @@ func (i *singleLevelIterator) firstInternal() (*InternalKey, base.LazyValue) {
 // SeekLT(upper))
 func (i *singleLevelIterator) Last() (*InternalKey, base.LazyValue) {
 	if i.vState != nil {
-		return i.virtualLast()
+		return i.maybeVerifyKey(i.virtualLast())
 	}
 
 	if i.upper != nil {

--- a/sstable/reader_iter_two_lvl.go
+++ b/sstable/reader_iter_two_lvl.go
@@ -617,6 +617,9 @@ func (i *twoLevelIterator) virtualLastSeekLE(key []byte) (*InternalKey, base.Laz
 		return nil, base.LazyValue{}
 	}
 	if result == loadBlockIrrelevant {
+		if i.lower != nil && i.cmp(ikey.UserKey, i.lower) < 0 {
+			i.exhaustedBounds = -1
+		}
 		// Load the previous block.
 		return i.skipBackward()
 	}

--- a/testdata/excise
+++ b/testdata/excise
@@ -431,3 +431,96 @@ prev
 cc: (., [cc-f) @7=foo UPDATED)
 bbsomethinglong@4: (bazz, . UPDATED)
 b@5: (bar, .)
+
+# Regression test for #3236. Lower bounds are enforced in virtualLastSeekLE()
+# in the case where the actual virtual sstable last key is invisible due to
+# being obsolete. We create an EFOS to ensure that we create a virtual
+# sstable consisting only of obsolete keys with an excise, then do an iterator
+# operation that does a virtualLastSeekLE() on it.
+
+reset
+----
+
+batch
+set a foo
+set b bar
+----
+
+batch
+set d@6 baz
+----
+
+flush
+----
+
+compact a z
+----
+
+batch
+set d@6 something
+----
+
+flush
+----
+
+batch
+set x something
+----
+
+file-only-snapshot s1
+a z
+----
+ok
+
+build ext7
+del d@6
+----
+
+ingest ext7
+----
+
+lsm
+----
+0.1:
+  000008:[d@6#15,DEL-d@6#15,DEL]
+0.0:
+  000007:[d@6#13,SET-d@6#13,SET]
+6:
+  000005:[a#10,SET-d@6#12,SET]
+
+compact c e
+----
+
+lsm
+----
+6:
+  000009:[a#0,SET-d@6#0,SET]
+
+build ext5
+set c something
+----
+
+ingest-and-excise ext5 excise=c-cc
+----
+
+lsm
+----
+6:
+  000011:[a#0,SET-b#0,SET]
+  000010:[c#16,SET-c#16,SET]
+  000012:[d@6#15,DEL-d@6#0,SET]
+
+iter lower=c upper=e
+last
+prev
+prev
+seek-lt dd
+prev
+prev
+----
+c: (something, .)
+.
+.
+c: (something, .)
+.
+.

